### PR TITLE
Multi-period and parallel PowerFlows

### PR DESCRIPTION
**Pull request by alefcastelli**
Multi-period implementation of PowerFlows. Parallel function were tested but not proposed. Might be in next commits if computational bottleneck is solved.

### CODE REVIEW

File: src/PowerFlowData.jl [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
+
+struct PowerFlowData{M <: PNM.PowerNetworkMatrix, N, S <: Union{String, Char}}
     bus_lookup::Dict{Int, Int}
     branch_lookup::Dict{String, Int}
-    bus_activepower_injection::Vector{Float64}
-    bus_reactivepower_injection::Vector{Float64}
-    bus_activepower_withdrawals::Vector{Float64}
-    bus_reactivepower_withdrawals::Vector{Float64}
+    bus_activepower_injection::Union{Vector{Float64}, Matrix{Float64}}
+    bus_reactivepower_injection::Union{Vector{Float64}, Matrix{Float64}}
```
Comment (jd-lara) [inline src/PowerFlowData.jl:8]: There is no need for this union. This can be just Matrix

File: src/PowerFlowData.jl [hunk 2]
```diff
@@ -1,20 +1,111 @@
...
+    if timesteps != 0
+        if length(timestep_names) == 0
+            timestep_names = [string(i) for i in 1:timesteps]
+        elseif length(timestep_names) != timesteps
+            error("timestep_names field must have same length as timesteps")
+        end
+    end
+
+    # get data for calculations
     power_network_matrix = PNM.Ybus(sys)
```
Comment (jd-lara) [inline src/PowerFlowData.jl:109]: We should add connectivity checks here

File: src/PowerFlowData.jl [hunk 3]
```diff
@@ -40,44 +131,114 @@ function PowerFlowData(::ACPowerFlow, sys::PSY.System)
         bus = PSY.get_component(PSY.Bus, sys, bus_name)
         bus_type[ix] = PSY.get_bustype(bus)
         bus_angles[ix] = PSY.get_angle(bus)
+        # ! shouldn't it be like this?
```
Comment (jd-lara) [inline src/PowerFlowData.jl:134]: Yes

File: src/PowerFlowData.jl [hunk 4]
```diff
@@ -40,44 +131,114 @@ function PowerFlowData(::ACPowerFlow, sys::PSY.System)
...
         bus_lookup,
         sys,
     )
 
+    # initialize data
+    init_1 = zeros(n_buses, timesteps)
+    init_2 = zeros(n_branches, timesteps)
+
+    # define fields as matrices whose number of columns is eqault to the number of timesteps
+    bus_activepower_injection_1 = deepcopy(init_1)
```
Comment (jd-lara) [inline src/PowerFlowData.jl:166]: using 

```suggestion
    bus_activepower_injection_1 = similar(init_1)
```

might be faster than copies

File: src/PowerFlowData.jl [hunk 5]
```diff
@@ -40,44 +131,114 @@ function PowerFlowData(::ACPowerFlow, sys::PSY.System)
...
+    bus_reactivepower_injection_1 = deepcopy(init_1)
+    bus_activepower_withdrawals_1 = deepcopy(init_1)
+    bus_reactivepower_withdrawals_1 = deepcopy(init_1)
+    bus_magnitude_1 = deepcopy(init_1)
+    bus_angles_1 = deepcopy(init_1)
+    branch_flow_values_1 = deepcopy(init_2)
+
+    # initial values related to first timestep allocated in the first column
+    bus_activepower_injection_1[:, 1] .= bus_activepower_injection
```
Comment (jd-lara) [inline src/PowerFlowData.jl:175]: why the `_1` nomenclature?

File: src/common.jl [hunk 6]
```diff
@@ -63,69 +63,92 @@ end
...
-)
-    y[:] .= transpose(A) * x
-    return
-end
-
-# virtual case: all lines
-
-function my_mul_mt!(
-    y::Vector{Float64},
+# # sparse case (ABA and BA)
```
Comment (jd-lara) [inline src/common.jl:66]: remove this code if it isn't going to be used.

File: src/post_processing.jl [hunk 7]
```diff
@@ -535,6 +521,90 @@ function write_results(
...
+    # get branches from/to buses
+    from_bus = Vector{Int}(undef, length(branches))
+    to_bus = Vector{Int}(undef, length(branches))
+    for (i, branch) in enumerate(branches)
+        br = PSY.get_component(PSY.ACBranch, sys, branch)
+        from_bus[i] = PSY.get_number(PSY.get_arc(br).from)
+        to_bus[i] = PSY.get_number(PSY.get_arc(br).to)
+    end
+
+    # if length(data.timestep_map) == 1
```
Comment (jd-lara) [inline src/post_processing.jl:555]: why is this commented out?

File: src/post_processing.jl [hunk 8]
```diff
@@ -535,6 +521,90 @@ function write_results(
...
+    # end
+    # return
+end
+
+# TODO: multi-period still to implement
+"""
+Returns a dictionary containing the AC power flow results.
+
+WARNING: single-period evalution only is suppoerted at the moment. Resulting
+dictionary will therefore feature just one key linked to one DataFrame.
```
Comment (jd-lara) [inline src/post_processing.jl:598]:
```suggestion
Only single-period evaluation is supported at the moment for AC Power flows. Resulting
dictionary will therefore feature just one key linked to one DataFrame.
```

File: src/solve_dc_powerflow.jl [hunk 9]
```diff
@@ -114,3 +201,70 @@ function solve_powerflow(
...
+Each DataFrame containts the flows and angles.
+
+# Arguments:
+- `data::PTDFPowerFlowData`:
+        PowerFlowData structure containing the system's data per each timestep
+        considered, as well as the PTDF matrix.
+- `sys::PSY.System`:
+        container gathering the system data.
+"""
+function solve_powerflow(
```
Comment (jd-lara) [inline src/solve_dc_powerflow.jl:220]:
```suggestion
function solve_powerflow!(
```

File: src/solve_dc_powerflow.jl [hunk 10]
```diff
@@ -114,3 +201,70 @@ function solve_powerflow(
...
+    data::PTDFPowerFlowData,
+    sys::PSY.System;
+)
+    solve_powerflow!(data)
+    return write_results(data, sys)
```
Comment (jd-lara) [inline src/solve_dc_powerflow.jl:225]:
```suggestion
    return write_results!(data, sys)
```

File: src/solve_dc_powerflow.jl [hunk 11]
```diff
@@ -114,3 +201,70 @@ function solve_powerflow(
...
+Each DataFrame containts the flows and angles.
+
+# Arguments:
+- `data::ABAPowerFlowData`:
+        PowerFlowData structure containing the system's data per each timestep
+        considered, as well as the ABA and BA matrices.
+- `sys::PSY.System`:
+        container gathering the system data.
+"""
+function solve_powerflow(
```
Comment (jd-lara) [inline src/solve_dc_powerflow.jl:242]:
```suggestion
function solve_powerflow!(
```

File: test/test/Project.toml
Comment (jd-lara) [inline test/test/Project.toml:1]: What is this file about?

File: test/test_powerflow_data.jl [hunk 12]
```diff
@@ -5,3 +5,12 @@
     PowerFlowData(PTDFDCPowerFlow(), sys)
     PowerFlowData(vPTDFDCPowerFlow(), sys)
 end
+
+@testset "PowerFlowData multiperiod" begin
+    sys = PSB.build_system(PSB.PSITestSystems, "c_sys14"; add_forecasts = false)
+    timesteps = 24
+    @warn "multiperiod AC still to implement"
```
Comment (jd-lara) [inline test/test_powerflow_data.jl:12]: we don't need to log a warning for this

```suggestion
# TODO: "multiperiod AC still to implement"
```

**Review by jd-lara — CHANGES_REQUESTED** [review]
The PR still needs some improvements

File: src/PowerFlowData.jl (hunk 1 above)
Comment (alefcastelli) [reply src/PowerFlowData.jl]: solved

File: src/PowerFlowData.jl (hunk 4 above)
Comment (alefcastelli) [reply src/PowerFlowData.jl]: done

File: src/PowerFlowData.jl (hunk 5 above)
Comment (alefcastelli) [reply src/PowerFlowData.jl:175]: to differentiate with respect to the vector of initial values (e.g., bus_activepower_injection and bus_activepower_injection_1)

File: src/post_processing.jl (hunk 7 above)
Comment (alefcastelli) [reply src/post_processing.jl:555]: I forgot the remove that. Now it is done.

File: src/PowerFlowData.jl (hunk 3 above)
Comment (alefcastelli) [reply src/PowerFlowData.jl:134]: fixed

File: src/common.jl (hunk 6 above)
Comment (alefcastelli) [reply src/common.jl]: done

File: test/test/Project.toml
Comment (alefcastelli) [reply test/test/Project.toml:1]: removed

File: src/solve_dc_powerflow.jl (hunk 9 above)
Comment (alefcastelli) [reply src/solve_dc_powerflow.jl:220]: I have defined two function, one that overwrites the input (solve_powerflows!), and one that returns a dictionary of data frames (solve_powerflows). It is true though that "solve_powerflows" calls "solve_powerflows!".

shouldn't be better to name the function writing a dictionary of dataframes with no "!" ?

File: src/solve_dc_powerflow.jl (hunk 10 above)
Comment (alefcastelli) [reply src/solve_dc_powerflow.jl:225]: same as above

File: src/solve_dc_powerflow.jl (hunk 11 above)
Comment (alefcastelli) [reply src/solve_dc_powerflow.jl:242]: same as above

File: src/PowerFlowData.jl (hunk 2 above)
Comment (alefcastelli) [reply src/PowerFlowData.jl:109]: This is how I found it. Do you want to add the connectivity check as for the DC case with the ABA and BA matrix? Or do you refer to something different?

**Pull request merged**